hail/python/hailtop/aiotools/validators.py

```diff
@@ -0,0 +1,63 @@
+from hailtop.aiotools.router_fs import RouterAsyncFS
+from hailtop.aiotools.utils import get_or_init_attr
+from typing import List
+from urllib.parse import urlparse
+
+
+class FileValidator:
+    """
+    A mixin that expects to have a ``hailtop.aiotools.router_fs.RouterAsyncFS`` available through the
+    ``router_async_fs`` property in order to call its ``validate_cloud_storage_policy`` method.
+    """
+    @property
+    def file_scheme_cache(self) -> List[str]:
+        """
+        Returns
+        -------
+        The list of URIs that have already had their scheme validated by ``.FileValidator.validate_file_scheme``.
+        If the cache doesn't already exist, initializes it with the empty list.
+        """
+        return get_or_init_attr(self, "_file_scheme_cache", [])
+
+    @property
+    def router_async_fs(self) -> RouterAsyncFS:
+        """
+        Returns
+        -------
+        The ``hailtop.aiotools.router_fs.RouterAsyncFS`` to which this instance has access.
+
+        Raises
+        ------
+        :class:`NotImplementedError`
+            If this instance does not have access to a ``hailtop.aiotools.router_fs.RouterAsyncFS``.
+        """
+        raise NotImplementedError
+
+    def validate_file(self, uri: str) -> None:
+        """
+        Validates the URI's scheme using ``.FileValidator.validate_file_scheme``, and delegates further file
+        validation to the ``hailtop.aiotools.router_fs.RouterAsyncFS``.
+        """
+        try:
+            self.validate_file_scheme(uri)
+            self.router_async_fs.validate_cloud_storage_policy(uri)
+        except NotImplementedError:
+            pass
+
+    def validate_file_scheme(self, uri: str) -> None:
+        """
+        By default, rejects any local filepaths. Override this method in subclasses to change that behavior.
+
+        Raises
+        ------
+        :class:`ValueError`
+            If the URI refers to a file on the local filesystem.
+        """
+        if uri not in self.file_scheme_cache:
+            scheme = urlparse(uri).scheme
+            if not scheme or scheme == "file":
+                raise ValueError(
+                    f"Local filepath detected: '{uri}'. The Hail Batch Service does not support the use of local filepaths. "
```

Comment:
wait, i'm confused, why is it not specific to the hail batch service? afaik the local and spark backends are allowed to use local filepaths, so this error only comes up when using QOB or the batch servicebackend, both of which ultimately use the batch service, no?

Reply:
You're right, the error does only get invoked in the correct contexts. But I think I phrased that comment poorly because that was not where my gripe was. I'll rephrase

This method acts basically like that of an abstract class, preventing repeating some code between `hailtop.batch.ServiceBackend` and `hail.backend.ServiceBackend`. It feels weird to me to have the method of the abstract class be "speaking" on behalf of a subset of subclasses, forcing other subclasses to override it if they don't want that behavior, which feels like a footgun. IMO it is better that `validate_file_scheme` be a method the local / spark backends are forced to implement as part of the type system than something we need to remember to override because we don't want the behavior of the abstract class.

Reply:
changed it so that backend subclasses have to have an implementation for `validate_file`, which can just call the reusable stateless function `hailtop.aiotools.validators.validate_file`, which accepts an optional argument for a `file_scheme_cache` and uses the presence of that arg to decide whether to check the file scheme